File: src/utils/laneMixer.js
/**
 * Lane Mixer v3.0 — MirrorDNA Lite Edition
 * More conservative with questions. Presence over interrogation.
 */

// ═══════════════════════════════════════════════════════════════
// BASELINE LANE MIX BY INTENT
// ═══════════════════════════════════════════════════════════════

// v3.0: More conservative mirror baselines
const BASELINE_MIRROR = {
    0: 0.00,   // Utility → 100% Direct, no questions
    1: 0.10,   // Info → 90% Direct, rarely ask
    2: 0.25,   // Choice → 75% Direct, sometimes reflect
    3: 0.50    // Personal → 50/50, reflect when meaningful
};

// ═══════════════════════════════════════════════════════════════
// DIAL EFFECT
// ═══════════════════════════════════════════════════════════════

function clamp(value, min, max) {
    return Math.min(Math.max(value, min), max);
}

/**
 * Compute effective lane mix from intent score and dial
 * v3.0: Dial has reduced effect (±15% instead of ±20%)
 */
export function computeLaneMix(intentScore, dial = 0.5) {
    const M_base = BASELINE_MIRROR[intentScore] ?? BASELINE_MIRROR[1];

    // v3.0: Reduced dial effect (±15%)
    const dialEffect = (dial - 0.5) * 0.3;

    // v3.0: Lower mirror cap (70% max)
    const mirror = clamp(M_base + dialEffect, 0, 0.70);
    const direct = 1 - mirror;

    return { direct, mirror };
}

// ═══════════════════════════════════════════════════════════════
// QUESTION CAPS — v3.0 CONSERVATIVE
// ═══════════════════════════════════════════════════════════════

/**
 * Get maximum questions allowed
 * v3.0: Much more conservative — questions are rare, not default
 */
export function getMaxQuestions(dial, intentScore) {
    // Utility (0) and Info (1) → NO questions unless dial is very high
    if (intentScore <= 1) {
        return dial > 0.8 ? 1 : 0;
    }

    // Choice (2) → 1 question only if dial is above center
    if (intentScore === 2) {
        return dial > 0.5 ? 1 : 0;
    }

    // Personal (3) → 1 question allowed, but model decides via mirror.present
    if (intentScore === 3) {
        return 1;
    }

    return 0;
}

// ═══════════════════════════════════════════════════════════════
// RENDERING DECISION
// ═══════════════════════════════════════════════════════════════

/**
 * Determine what to render based on lane mix
 * v3.0: Always show direct, mirror only when significant
 */
export function getRenderStrategy(laneMix) {
    const showDirect = true;  // v3.0: Always show direct
    const showMirror = laneMix.mirror > 0.15;  // v3.0: Higher threshold
    const directFirst = true;  // v3.0: Always lead with direct

    return { showDirect, showMirror, directFirst };
}

// ═══════════════════════════════════════════════════════════════
// FULL PIPELINE
// ═══════════════════════════════════════════════════════════════

/**
 * Compute full lane configuration
 */
export function computeLaneConfig(intentScore, dial = 0.5) {
    const laneMix = computeLaneMix(intentScore, dial);
    const maxQuestions = getMaxQuestions(dial, intentScore);
    const renderStrategy = getRenderStrategy(laneMix);

    return {
        intentScore,
        dial,
        laneMix,
        maxQuestions,
        ...renderStrategy
    };
}

File: mirror_gate/core/lane_mixer.py
import random
from typing import Dict, Any, List

class LaneMixer:
    def __init__(self):
        self.baseline_mirror = {0: 0.0, 1: 0.15, 2: 0.40, 3: 0.70}
        
        self.assumptions_templates = [
            "You may be assuming {0}, and {1}.", 
            "This rests on {0} — and {1}."
        ]
        self.question_templates = [
            "⟡ {question}", 
            "{question}"
        ]

    def compute_lane_mix(self, intent_score: int, dial: float = 0.5) -> Dict[str, float]:
        m_base = self.baseline_mirror.get(intent_score, 0.15)
        dial_effect = (dial - 0.5) * 0.4
        mirror = max(0, min(0.85, m_base + dial_effect))
        return {"direct": 1 - mirror, "mirror": mirror}

    def get_max_questions(self, dial: float, intent_score: int) -> int:
        if intent_score == 0 and dial <= 0.66: return 0
        if dial <= 0.33: return 1
        if dial <= 0.66: return 1
        if dial > 0.66 and intent_score >= 2: return 2
        return 1

    def render_two_lane(self, schema: Dict[str, Any], lane_mix: Dict[str, float], intent_score: int, max_questions: int) -> str:
        parts = []
        show_direct = lane_mix["direct"] > 0.1
        show_mirror = lane_mix["mirror"] > 0.1
        
        # Direct Lane
        if show_direct and schema.get("direct", {}).get("content"):
            parts.append(schema["direct"]["content"])
        
        # Mirror Lane
        if show_mirror and schema.get("mirror"):
            mirror = schema["mirror"]
            
            if mirror.get("assumptions") and isinstance(mirror["assumptions"], list):
                a = mirror["assumptions"]
                if len(a) > 0:
                    txt = random.choice(self.assumptions_templates)
                    txt = txt.replace("{0}", a[0] if len(a) > 0 else "something")
                    txt = txt.replace("{1}", a[1] if len(a) > 1 else "something else")
                    parts.append(txt)
            
            if max_questions > 0 and mirror.get("question"):
                q_text = random.choice(self.question_templates).format(question=mirror["question"])
                parts.append(q_text)
        
        return "\n\n".join(parts)
